toc.md

```diff
@@ -428,77 +429,232 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+
+When building a BJP it is important to see how the current implementation compares to a leader in the field. We decided to benchmark Workerholic against sidekiq in order to get some context as to where Workerholic is at and if we could optimize its performance.
 
 #### Serialization
+
+On our first iteration, we found that there was a great difference between Workerholic and Sidekiq; the former was much slower on both the enqueuing and the processing sides:
+
 ![optimizations_serialization_benchmark_yaml](/images/optimizations_serialization_benchmark_yaml.png)
 
-On our first iteration, we found that there was a great difference between Workerholic and Sidekiq; ours took much longer both on the enqueuing side and the processing side. Why was that? We looked into Sidekiq and found that it was using JSON serialization while we were using YAML, and so we decided to change our code to use JSON and see if that was really where the bottleneck was.
+What was our bottleneck? Sidekiq uses a similar model to Workerholic: concurrency via the use of threads. The difference seemed too big, especially since Workerholic should be a lighter BJP with less features and edge cases covered.
+
+The first thing we noticed is that both the enqueuing side and processing side were slower. This gave us a major insight into realizing it was a bottleneck present on both end of the spectrum. We started wondering if the way we serialize our data may be the bottleneck. Indeed, we were using YAML while Sidekiq was using JSON.
+So we decided to switch to a JSON serialization. The following YAML serialization:
+
+```ruby
+module Workerholic
+  class JobSerializer
+    # ...
+    def self.serialize(job)
+      YAML.dump(job.to_hash)
+    end
+
+    def self.deserialize(yaml_job)
+      job_info = YAML.load(yaml_job)
+      JobWrapper.new(job_info)
+    end
+
+    # ...
+  end
+end
+```
+
+Turned into the following JSON serialization:
+
+```ruby
+module Workerholic
+  class JobSerializer
+    # ...
+
+    def self.serialize(job)
+      JSON.dump(job.to_hash)
+    end
+
+    def self.deserialize(json_job)
+      job_info = JSON.parse(json_job, symbolize_names: true)
+
+      job_info[:klass] = job_info[:klass] ? Object.const_get(job_info[:klass]) : nil
+      job_info[:wrapper] = job_info[:wrapper] ? Object.const_get(job_info[:wrapper]) : nil
+
+      JobWrapper.new(job_info)
+    end
+
+    # ...
+  end
+end
+```
+
+After switching to JSON we benchmarked Workerholic against Sidekiq one more time:
 
 ![optimizations_serialization_benchmark_json](/images/optimizations_serialization_benchmark_json.png)
 
-And voila! Here are the results above. By changing our serialization strategy to JSON, we were able to improve our enqueuing duration by 70% and our processing duration by 55% compared to our YAML iteration. We're now slightly faster than Sidekiq!
+We were able to improve our enqueuing duration by 70% and our processing duration by 55% compared to our YAML iteration!
```

Comment:
We were able to reduce our enqueuing time by 70% and processing time by 55% compared to our previous YAML iteration!